delocate/delocating.py

```diff
@@ -578,6 +591,191 @@ def _make_install_name_ids_unique(
         validate_signature(lib)
 
 
+def _get_macos_min_version(dylib_path: Path) -> Iterable[Tuple[str, Version]]:
+    """Get the minimum macOS version from a dylib file.
+
+    Parameters
+    ----------
+    dylib_path : Path
+        The path to the dylib file.
+
+    Returns
+    -------
+    List[Tuple[str, Version]]
+        A list of tuples containing the CPU type and the minimum macOS version.
+    """
+    m = MachO(dylib_path)
+    res = []
+    for header in m.headers:
+        for cmd in header.commands:
+            if cmd[0].cmd == LC_BUILD_VERSION:
+                version = cmd[1].minos
+            elif cmd[0].cmd == LC_VERSION_MIN_MACOSX:
+                version = cmd[1].version
+            else:
+                continue
+            res.append(
+                (
+                    CPU_TYPE_NAMES.get(header.header.cputype, "unknown"),
+                    Version(f"{version >> 16 & 0xFF}.{version >> 8 & 0xFF}"),
+                )
+            )
+            break
+    return res
+
+
+def _get_archs_and_version_from_wheel_name(
+    wheel_name: str,
+) -> Dict[str, Version]:
+    """
+    Get the architecture and minimum macOS version from the wheel name.
+
+    Parameters
+    ----------
+    wheel_name : str
+        The name of the wheel.
+
+    Returns
+    -------
+    Dict[str, Version]
+        A dictionary containing the architecture and minimum macOS version
+        for each architecture in the wheel name.
+    """
+    platform_tag_set = parse_wheel_filename(wheel_name)[-1]
+    res = {}
+    for platform_tag in platform_tag_set:
+        match = _PLATFORM_REGEXP.match(platform_tag.platform)
+        if match is None:
+            raise ValueError(f"Invalid platform tag: {platform_tag.platform}")
+        major, minor, arch = match.groups()
+        res[arch] = Version(f"{major}.{minor}")
+    return res
+
+
+def _get_problematic_libs(
+    version: Version, version_lib_dict: Dict[Version, List[Path]]
+) -> Set[Path]:
+    """
+    Filter libraries that require more modern macOS
+    version than the provided one.
+
+    Parameters
+    ----------
+    version : Version
+        The expected minimum macOS version
+    version_lib_dict : Dict[Version, List[Path]]
+        A dictionary containing mapping from minimum macOS version to libraries
+        that require that version.
+
+    Returns
+    -------
+    Set[Path]
+        A set of libraries that require a more modern macOS version than the
+        provided one.
+    """
+    result = set()
+    for v, lib in version_lib_dict.items():
+        if v > version:
+            result.update(lib)
+    return result
+
+
+def _calculate_minimum_wheel_name(
+    wheel_name: str, wheel_dir: Path
+) -> Tuple[str, set[Path]]:
```

Comment:
Replace the Python 2 `from __future__ import division, print_function` with `from __future__ import annotations`. This will let you use PEP 585 generics in type-hints.